addons/packages/contour/1.20.0/bundle/config/contour.star

```diff
@@ -0,0 +1,120 @@
+    return "Cluster"
+  else:
+    return "Local"
+  end
+end
+
+def get_envoy_service_annotations():
+  annotations = {}
+
+  if data.values.infrastructureProvider == "aws":
+    if data.values.envoy.service.aws.loadBalancerType == "nlb":
+      annotations["service.beta.kubernetes.io/aws-load-balancer-type"] = "nlb"
+    else:
+      annotations["service.beta.kubernetes.io/aws-load-balancer-backend-protocol"] = "tcp"
+      if data.values.contour.useProxyProtocol:
+        annotations["service.beta.kubernetes.io/aws-load-balancer-proxy-protocol"] = "*"
+      end
+    end
+  end
+
+  if data.values.envoy.service.annotations:
+    annotations_kvs = struct.decode(data.values.envoy.service.annotations)
+    annotations.update(annotations_kvs)
+  end
+
+  return annotations
+end
+
+# ##########
+# VALIDATION
+# ##########
+
+def validate_contour():
+  validate_funcs = [validate_infrastructure_provider,
+                    validate_contour_namespace,
+                    validate_contour_deployment,
+                    validate_contour_certificate,
+                    validate_envoy_deployment,
+                    validate_envoy_service]
+   for validate_func in validate_funcs:
+     validate_func()
+   end
+end
+
+def validate_infrastructure_provider():
+  if data.values.infrastructureProvider:
+    data.values.infrastructureProvider in ("docker", "aws", "vsphere", "azure") or assert.fail("infrastructureProvider must be either docker or aws or vsphere or azure")
+  end
+end
+
+def validate_contour_namespace():
+  data.values.namespace or assert.fail("namespace must be provided")
+end
+
+def validate_contour_deployment():
+  data.values.contour.replicas or assert.fail("contour.replicas must be provided")
+end
+
+def validate_contour_certificate():
+  if data.values.certificates.useCertManager:
+    data.values.certificates.duration or assert.fail("certificates.duration must be provided when certificates.useCertManager is true")
+    data.values.certificates.renewBefore or assert.fail("certificates.renewBefore must be provided when certificates.useCertManager is true")
+  end
+end
+
+def validate_envoy_deployment():
+  if data.values.envoy.hostPorts.enable:
+    data.values.envoy.hostPorts.http or assert.fail("envoy.hostPorts.http must be provided when envoy.hostPorts.enable is true")
+    data.values.envoy.hostPorts.https or assert.fail("envoy.hostPorts.https must be provided when envoy.hostPorts.enable is true")
+  end
+  
+  data.values.envoy.logLevel in ("trace", "debug", "info", "warning", "warn", "error", "critical", "off") or assert.fail("envoy.logLevel must be one of trace|debug|info|warning/warn|error|critical|off")
+  
+  data.values.envoy.terminationGracePeriodSeconds or assert.fail("envoy.terminationGracePeriodSeconds must be provided")
+end
+
+def validate_envoy_service():
+  if data.values.envoy.service.type:
+    # TODO should we relax this constraint? Not sure why we need to disallow ClusterIP for vsphere.
```

Comment:
I'm leaning towards removing this restriction of no ClusterIP services for vsphere. I'm not aware of why it's there, and I feel like a user will bump into it at some point and we'll have no good reason not to allow it. It still defaults to "NodePort" for vsphere, so only will use ClusterIP if the user explicitly specifies it.

Reply:
Dropped this restriction.